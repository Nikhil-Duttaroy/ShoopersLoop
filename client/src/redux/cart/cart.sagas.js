import { all, call, takeLatest, put, select } from "redux-saga/effects";

import UserActionTypes from "../user/user.types";
import { addItem, clearCart, setCartFromFirebase } from "./cart.actions";
import CartActionTypes from './cart.types';
import { selectCartId, selectCartItems } from './cart.selector';
import { getUserCartRef } from "../../firebase/firebase.utils";
import { selectCurrentUser } from './../user/user.selector';

export function* clearCartOnSignOut() {
  yield put(clearCart());
}

export function* clearCartItemsOnCheckout() {
  yield put(clearCart());
}

// export function* updateCartInFirebase() {
//   const currentUser = yield select(selectCurrentUser);
//   if (currentUser) {
//     try {
//       const cartRef = yield getUserCartRef(currentUser.id);
//       const cartItems = yield select(selectCartItems);
//       yield cartRef.update({ cartItems });
//     } catch (error) {
//       alert(error);
//     }
//   }
// }
export function* updateCartInFirebase() {
  const currentUser = yield select(selectCurrentUser);

  if (currentUser) {
    try {
      const cartRef = yield getUserCartRef(currentUser.id);
      const cartItems = yield select(selectCartItems);
      const currentCartId = yield select(selectCartId);
      yield cartRef.set({
        id: currentCartId,
        cartItems,
      });
    } catch (error) {
      console.log(error);
    }
  }
}
// export function* checkCartFromFirebase({ payload: user }) {
//   const cartRef = yield getUserCartRef(user.id);
//   const cartSnapshot = yield cartRef.get();
//   yield put(setCartFromFirebase(cartSnapshot.data().cartItems));
// }

export function* mergeCartWithFirebase() {
  const currentUser = yield select(selectCurrentUser);
  const currentCartId = yield select(selectCartId);

  try {
    const loggedOutCartItems = yield select(selectCartItems);
    const cartRef = yield getUserCartRef(currentUser.id);
    const cartSnapshot = yield cartRef.get();
    const cartData = yield cartSnapshot.data();

    yield put(setCartFromFirebase(cartData.cartItems));

    if (currentCartId !== cartData.id) {
      for (let i = 0; i < loggedOutCartItems.length; i++) {
        yield put(addItem(loggedOutCartItems[i]));
      }
    }
  } catch (error) {
    console.log(error);
  }
}

export function* onSignOutSuccess() {
  yield takeLatest(UserActionTypes.SIGN_OUT_SUCCESS, clearCartOnSignOut);
}
export function* onCheckoutSuccess() {
  yield takeLatest(CartActionTypes.CHECKOUT_SUCCESS, clearCartItemsOnCheckout);
}

export function* onUserSignIn() {
  yield takeLatest(UserActionTypes.SIGN_IN_SUCCESS,  mergeCartWithFirebase);
}

export function* onCartChange() {
  yield takeLatest(
    [
      CartActionTypes.ADD_ITEM,  
      CartActionTypes.REMOVE_ITEM,
      CartActionTypes.CLEAR_ITEM_FROM_CART,
    ],
    updateCartInFirebase
  );
}

export function* cartSagas() {
  yield all([call(onSignOutSuccess), call(onCartChange), call(onUserSignIn),call(onCheckoutSuccess)]);
  // yield all([call(onSignOutSuccess)],[call(onCheckoutSuccess)],[call(onCartChange)], [call(onUserSignIn)]);
}
