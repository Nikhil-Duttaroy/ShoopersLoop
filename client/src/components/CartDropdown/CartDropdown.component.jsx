import React from "react";
import { connect,useSelector} from "react-redux";
import CartItem from "./../CartItem/CartItem.component";

import { createStructuredSelector } from "reselect";
import { withRouter } from "react-router-dom";

import { selectCartItems } from "./../../redux/cart/cart.selector";
import { toggleCartHidden } from "../../redux/cart/cart.actions";



import {
  CartDropdownContainer,
  CartDropdownButton,
  EmptyMessageContainer,
  CartItemsContainer,
} from "./CartDropdown.styles";

export const CartDropdown = ({ cartItems, history, dispatch }) => {
  const user=useSelector(state=>state.user.currentUser)
  // console.log(user);
  return (
    <CartDropdownContainer>
      <CartItemsContainer>
        {cartItems.length ? (
          cartItems.map((cartItem) => (
            <CartItem key={cartItem.id} item={cartItem} />
          ))
        ) : (
          <EmptyMessageContainer>Your cart is empty</EmptyMessageContainer>
        )}
      </CartItemsContainer>
      <CartDropdownButton
        onClick={() => {
          user ? history.push("/checkout") : history.push("/signin");
          dispatch(toggleCartHidden());
        }}
      >
        {user ? "GO TO CHECKOUT" : "SignIN"}
      </CartDropdownButton>
    </CartDropdownContainer>
  );  
};

const mapStateToProps = createStructuredSelector({
  cartItems: selectCartItems,
});

export default withRouter(connect(mapStateToProps)(CartDropdown));
