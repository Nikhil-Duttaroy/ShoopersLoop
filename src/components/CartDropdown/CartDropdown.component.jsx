import React from 'react'
import { connect } from "react-redux";
import CustomButton from './../CustomButton/CustomButton.components';
import CartItem from "./../CartItem/CartItem.component";

import { selectCartItems } from "./../../redux/cart/cart.selector";

import './CartDropdown.styles.scss'



const CartDropdown = ({ cartItems }) => (
  <div className='cart-dropdown'>
    <div className='cart-items'>
      {cartItems.map(cartItem => (
        <CartItem key={cartItem.id} item={cartItem}/>
      ))}
    </div>
    <CustomButton>GO TO CHECKOUT</CustomButton>
  </div>
);

//*Removed as we are using reselect library
// const mapStateToProps =({cart:{cartItems}}) => ({
//   cartItems
// })

const mapStateToProps =(state) => ({
   cartItems:selectCartItems(state)
})

export default connect(mapStateToProps)(CartDropdown);