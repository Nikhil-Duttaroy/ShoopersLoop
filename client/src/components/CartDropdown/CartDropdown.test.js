import React from "react";
import { shallow } from "enzyme";

import CartItem from "../CartItem/CartItem.component";

import { toggleCartHidden } from "../../redux/cart/cart.actions";
import {CartDropdown} from './CartDropdown.component';

import * as redux from "react-redux";


describe("CartDropdown component", () => {
  let wrapper;
  let mockHistory;
  let mockDispatch;
  const mockCartItems = [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }, { id: 5 }];

    const spy = jest.spyOn(redux, "useSelector");
    spy.mockReturnValue({ user: "test" });
    

  beforeEach(() => {
    mockHistory = {
      push: jest.fn(),
    };

    mockDispatch = jest.fn();

    const mockProps = {
      cartItems: mockCartItems,
      history: mockHistory,
      dispatch: mockDispatch,
    };

    wrapper = shallow(<CartDropdown {...mockProps} />);
  });

  it("should render CartDropdown component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should call history.push when button is clicked", () => {
    wrapper.find("CartDropdownButton").simulate("click");
    expect(mockHistory.push).toHaveBeenCalled();
    expect(mockDispatch).toHaveBeenCalledWith(toggleCartHidden());
  });

  it("should render an equal number of CartItem components as the cartItems prop", () => {
    expect(wrapper.find(CartItem).length).toEqual(mockCartItems.length);
  });

  it("should render EmptyMessageContainer if cartItems is empty", () => {
    const mockProps = {
      cartItems: [],
      history: mockHistory,
      dispatch: mockDispatch,
    };

    const newWrapper = shallow(<CartDropdown {...mockProps} />);
    expect(newWrapper.exists("EmptyMessageContainer")).toBe(true);
  });
});
