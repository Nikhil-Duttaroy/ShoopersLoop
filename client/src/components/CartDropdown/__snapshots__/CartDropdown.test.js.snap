// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartDropdown component should render CartDropdown component 1`] = `
<styled.div>
  <styled.div>
    <Memo(CartItem)
      item={
        Object {
          "id": 1,
        }
      }
      key="1"
    />
    <Memo(CartItem)
      item={
        Object {
          "id": 2,
        }
      }
      key="2"
    />
    <Memo(CartItem)
      item={
        Object {
          "id": 3,
        }
      }
      key="3"
    />
    <Memo(CartItem)
      item={
        Object {
          "id": 4,
        }
      }
      key="4"
    />
    <Memo(CartItem)
      item={
        Object {
          "id": 5,
        }
      }
      key="5"
    />
  </styled.div>
  <CartDropdownButton
    onClick={[Function]}
  >
    SignIN
  </CartDropdownButton>
</styled.div>
`;
